File: chapter5/startup.js
/*
 * Server launcher
 */
var fs = require ("fs");
var fork = require ("child_process").fork;

//if module invoked directly, call main
if (require.main === module) {
    main ();
}

/**
 * Load configuration and initiate server launches.
 */
function main () {
    console.log ("Reading configuration from startup.json.");
    // load configuration file, with server startup data
    var configuration = null;
    try {
        var data = fs.readFileSync ("startup.json");
        configuration = JSON.parse (data);
    } catch (e) {
        console.log ("Error loading configuration: " + e.message);
        process.exit (1);
    }
    // start servers
    launchServers (configuration.servers);
}

/**
 * Launch all servers marked with start:true.
 * @param {object} servers List of servers to launch
 */
function launchServers (servers) {
    // for each server in configuration
    for (var ctr = 0; ctr < servers.length; ctr++) {
        var server = servers[ctr];
        // if server marked to start
        if (server.start) {
            // populate port number in args, start child process
            console.log ("Starting " + server.name);
            var args = ["--port=" + server.port];
            fork (server.program + ".js", args, null, startError);
        }
    }
}

/**
 * On error starting a server, display message, terminate program.
 * @param {string} error Error message
 */
function startError (error) {
    console.log ("Error starting server: " + error);
    process.exit (1);
}

File: chapter5/startup.py
"""
Server launcher
"""
from json import loads
from subprocess import Popen
import sys
from time import sleep

def main ():
    print ("Reading configuration from startup.json.")

    # load configuration file, which contains the startup
    # directions for all servers.
    try:
        # read the file and convert to a JSON object
        data = open ("startup.json", "rU").read ()
    except IOError as e:
        print ("Error reading configuration file: " + e.strerror)
        sys.exit (1)

    try:
        configuration = loads (data)
    except Exception as e:
        print ("Invalid JSON content in startup.json")
        sys.exit (1)

    launchServers (configuration["servers"])

def launchServers (servers):
    # for each server in configuration
    processes = []
    for server in servers:
        # if server marked to start
        if server["start"]:
            # set port number in arguments and start child process
            print ("Starting " + server["name"])
            program = server["program"] + ".py"
            port = "--port=" + str (server["port"])
            process = Popen (["python", program, port])
            processes.append (process)
            # allow child process messages to display
            sleep (0.25)

    for process in processes:
        process.wait ()

if __name__ == "__main__":
    main ()